### ISSUE TITLE
Impersonation (Auth Hook JWT) does not reflect production behaviour

### ISSUE THREAD

**Issue by activenode** [supabase/supabase#27841]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

When using the Impersonation feature of Supabase Studio, it doesn't reflect the actual permissions as it does not call the defined **Custom Access Token Hook**.

This in fact is not a feature request but a bug because Supabase Studio by all means must reflect production behaviour when it comes to impersonation with regards to the authenticated user.

The original discussion is found here: https://github.com/orgs/supabase/discussions/22482#discussioncomment-9011500 

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Create a Custom Access Token Hook as follows:

```
CREATE OR REPLACE FUNCTION public.custom_access_token_hook(event jsonb)
RETURNS jsonb
LANGUAGE plpgsql
AS $$
  DECLARE
    claims jsonb;
    user_id uuid;
  BEGIN
    user_id := (event->>'user_id')::uuid;

    claims := event->'claims';

    -- Check if 'app_metadata' exists in claims
    if jsonb_typeof(claims->'app_metadata') is null then
        -- If 'app_metadata' does not exist, create an empty object
        claims := jsonb_set(claims, '{app_metadata}', '{}');
    end if;


    -- Set a claim of 'admin'
    claims := jsonb_set(claims, '{app_metadata, combo}', '["snoo"]');

    -- Update the 'claims' object in the original event
    event := jsonb_set(event, '{claims}', claims);

    -- Return the modified or original event
    return event;
  END;
$$;

grant execute
  on function public.custom_access_token_hook
  to supabase_auth_admin;
```

2. Go to the SQL editor and impersonate _any_ authenticated user
3. Execute `SELECT auth.jwt();`
4. See `combo` property missing

## Expected behavior

With the given custom hook `combo: ["snoo"]` should show up in the user object of `SELECT auth.jwt()`.

**Comment by encima** [comment]
Thanks for opening @activenode, this does seem to be both a bug and a feature request. Passing to @supabase/frontend to decide on the best path forward here.

**Comment by ardabeyazoglu** [comment]
@encima I don't think this bug is only related to frontend or supabase studio. I have a similar problem with token hook not reflecting app_metadata to supabase client. I think the same issue. You can reproduce it easily.

1. Create a token hook like @activenode shared, set a new variable in app_metadata
2. Call `(await supabase.auth.refreshSession()).data` from client (or sign out, and sign in again)
3. Now you will see that, you don't have the new key/value in the user object. However, if you put the new access token to [jwt.io](https://jwt.io), you will find the new value in app_metadata. So, token hook is actually writing new data to the jwt, but getUser/refreshSession methods don't process it.

This is a barrier for using access token hooks for now.

Do you want me to create another issue for this, or keep it here?

**Comment by encima** [comment]
Hey @ardabeyazoglu 

Interesting, thank you for providing the steps for this!

For now, lets let the @supabase/auth team triage this as a single issue and determine if this needs to be split up.

**Comment by GaryAustin1** [comment]
@ardabeyazoglu  that is expected and called out in the docs.  You have to use a jwt library to decode the claims in the access_token as the auth-hooks do not change the auth.user meta data values which are returned in the user object.

**Comment by ardabeyazoglu** [comment]
@GaryAustin1 thanks, i see how it works. 
Updating raw_app_meta_data in the hook solves my problem for now.

**Comment by GaryAustin1** [comment]
@Hallidayo 
This is still a feature request/ bug that impacts lots of users.
Unless the team has it elsewhere as a needed improvement or documents make it clear it should stay open IMO.

**Comment by Hallidayo** [comment]
@GaryAustin1 - I'll re-open and flag it up.

**Comment by WolfspiritM** [comment]
Just stumbled over this when trying to debug RBAC rules following this documentation:
https://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac

I think it should at least be mentioned in the docs that impersonation via studio doesn't trigger the custom hook

**Comment by nadeemkhoury** [comment]
any update on this issue ?

**Comment by AdzeB** [comment]
@alaister Hi, does the PR support local studio or that's not averrable yet?

**Comment by alaister** [comment]
@AdzeB Hi, yes this should be available locally. Please make sure you're running the latest version of the CLI and dashboard in your local setup – you may need to upgrade to see this feature.

**Comment by sebahacket** [comment]
@alaister For some reason custom hooks are still not executing in our setup. We have upgraded. 

Using Supabase CLI 2.23.4 and studio image: public.ecr.aws/supabase/studio:2025.05.19-sha-3487831

**Comment by stefan-girlich** [comment]
Unfortunately this does not seem to be working as of Supabase CLI v2.30.4 . The output from `select auth.jwt()` does not contain the custom values added to `"app_metadata"`. This renders the impersonation feature entirely useless in my case. Custom values from my auth hook are correctly added in sessions outside of Supabase Studio.

@joshenlim Would you be able to provide support? Maybe there are additional steps required that I am missing.

**Comment by GaryAustin1** [comment]
> Unfortunately this does not seem to be working as of Supabase CLI v2.30.4 . The output from `select auth.jwt()` does not contain the custom values added to `"app_metadata"`. This renders the impersonation feature entirely useless in my case. Custom values from my auth hook are correctly added in sessions outside of Supabase Studio.
> 
> [@joshenlim](https://github.com/joshenlim) Would you be able to provide support? Maybe there are additional steps required that I am missing.

This is the way hosted also works.  The auth hook changes are not reflected in impersonation as it is reading the database to get the values out of auth.users.  There is no JWT auth call involved with getting the impersonation data so the hook is not run.

**Comment by activenode** [comment]
> Custom values from my auth hook are correctly added in sessions outside of Supabase Studio.

The hook is called hook because it is called on a specific event, which is signing in in your app. The `auth.jwt()` function however represents the value of the `jwt` at a certain moment.

> This renders the impersonation feature entirely useless in my case. 

I understand but what you're asking for actually might run tests useless where the jwt is not generated by the Auth Hook (e.g. via jwt secret).

So I'm not so sure if it would be the right thing to do that. If ever, as an optional feature.

**Comment by activenode** [comment]
Sorry for the wrong notifications here, I was confused by the thread. So long story short, why did we close it due to inactivity although still open? Could someone enlighten me?

**Comment by SebastianMFrank** [comment]
Will this be tackled at all in the future? The user impersonation feature is quite valuable for quick & easy verification but unfortunately useless in our case as well due to custom claims. Or is there a working solution that I missed?

**Comment by Korat-Dishant** [comment]
> > Unfortunately this does not seem to be working as of Supabase CLI v2.30.4 . The output from `select auth.jwt()` does not contain the custom values added to `"app_metadata"`. This renders the impersonation feature entirely useless in my case. Custom values from my auth hook are correctly added in sessions outside of Supabase Studio.
> > [@joshenlim](https://github.com/joshenlim) Would you be able to provide support? Maybe there are additional steps required that I am missing.
> 
> This is the way hosted also works. The auth hook changes are not reflected in impersonation as it is reading the database to get the values out of auth.users. There is no JWT auth call involved with getting the impersonation data so the hook is not run.


@GaryAustin1  I am able to see the user role when I run ```select (auth.jwt()  ->> 'user_role')```  in the production. but the same thing does not work in local development setup. Any idea why this is happening?

**Comment by PidgeyBE** [comment]
Same issue here... Can't debug production issue in development because custom JWT claims are missing... :/
